src/mem/cache/prefetch/ipcp.cc

```diff
@@ -0,0 +1,398 @@
+/*
+ * Copyright (c) 2024 Samsung Electronics
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * Describes a IPCP prefetcher based on template policies.
+ */
+
+#include "mem/cache/prefetch/ipcp.hh"
+
+#include "debug/HWPrefetch.hh"
+#include "params/IpcpPrefetcher.hh"
+
+namespace gem5
+{
+
+namespace prefetch
+{
+
+Ipcp::Ipcp(const IpcpPrefetcherParams &p)
+    : Queued(p), bitsIpTag(p.ip_tag_bits),
+    bitsLastVpage(p.last_vpage_bits),
+    bitsSignature(p.signature_bits),
+    degree(p.degree), csOn(p.cs_on),
+    cplxOn(p.cplx_on),
+    //GS IP Prefetcher
+    RM_SIZE(p.rm_size), REGION_SIZE(p.region_size),
+    pageDegree(p.page_degree), pageOn(p.page_on),
+    ipcpStats(this), nlOn(p.nl_on)
+{
+    bitsIpIndex = floorLog2(p.ip_entries);
+    bitsCsptIndex = floorLog2(p.cspt_entries);
+
+    std::cout << "bitsIpIndex: " << bitsIpIndex << std::endl;
+
+    ipcpTable.clear();
+    cspt.clear();
+
+    //GS IP Prefetcher
+    RST.clear();
+
+    misses = 0;
+    mpkc = 0;
+
+}
+
+Addr
+Ipcp::getSignature(Addr curSignature, int64_t stride)
+{
+    //Handle -ve stride
+    int64_t corStride = (stride < 0) ?
+     (stride * (-1)
+     + (1 << floorLog2(blkSize))) : stride;
+    return (((curSignature << 1) ^ corStride)
+     & ((1 << bitsSignature) - 1));
```

Comment:
I think if/else would be much easier to read! Especially since the ternary statement goes over so many lines.